File: sdks/rust/src/resources/license_keys.rs
//! License-key resource (`/v1/license-keys`).

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::{Client, LicenseKey, Result};

/// Accessor for license-key operations.
#[derive(Debug)]
pub struct LicenseKeys<'a> {
    client: &'a Client,
}

/// Body for issuing a license key (`POST /v1/license-keys`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct IssueLicenseKey {
    pub organization_id: String,
    pub customer_id: String,
    pub product_id: String,
    pub entitlement_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub machine_limit: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain_limit: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
}

/// Body for verifying a license key (`POST /v1/license-keys/verify`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VerifyLicenseKey {
    pub license_key: String,
    pub product_id: String,
    pub machine_id: String,
}

/// An offline validation token minted for a license key.
#[derive(Debug, Clone, Deserialize)]
pub struct LicenseToken {
    pub token: String,
}

impl<'a> LicenseKeys<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        LicenseKeys { client }
    }

    /// Issue a new license key.
    pub async fn issue(&self, body: &IssueLicenseKey) -> Result<LicenseKey> {
        self.client
            .request(Method::POST, "/v1/license-keys", Some(body))
            .await
    }

    /// Verify a license key for a product + machine. The deserialized result
    /// captures the API's verification payload.
    pub async fn verify(&self, body: &VerifyLicenseKey) -> Result<serde_json::Value> {
        self.client
            .request(Method::POST, "/v1/license-keys/verify", Some(body))
            .await
    }

    /// Mint an offline validation token for an existing license.
    pub async fn issue_token(&self, id: &str) -> Result<LicenseToken> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/license-keys/{id}/token"), None)
            .await
    }

    /// Revoke a license key.
    pub async fn revoke(&self, id: &str) -> Result<LicenseKey> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/license-keys/{id}/revoke"), None)
            .await
    }
}

File: sdks/python/settlekit/resources/license_keys.py
"""License keys resource — issue and verify machine-bound license keys."""

from __future__ import annotations

from typing import Any

from ._base import AsyncResource, SyncResource


def _create_body(
    *,
    organization_id: str,
    customer_id: str,
    product_id: str,
    entitlement_id: str,
    machine_limit: int,
) -> dict[str, Any]:
    return {
        "organizationId": organization_id,
        "customerId": customer_id,
        "productId": product_id,
        "entitlementId": entitlement_id,
        "machineLimit": machine_limit,
    }


class LicenseKeys(SyncResource):
    """Synchronous license-keys namespace."""

    def create(
        self,
        *,
        organization_id: str,
        customer_id: str,
        product_id: str,
        entitlement_id: str,
        machine_limit: int,
    ) -> Any:
        """Issue a license key. ``POST /v1/license-keys``."""
        body = _create_body(
            organization_id=organization_id,
            customer_id=customer_id,
            product_id=product_id,
            entitlement_id=entitlement_id,
            machine_limit=machine_limit,
        )
        return self._request("POST", "/v1/license-keys", body=body)

    def verify(self, *, license_key: str, product_id: str, machine_id: str) -> Any:
        """Verify a license key for a machine.

        ``POST /v1/license-keys/verify`` -> ``{"active": bool, ...}``.
        """
        return self._request(
            "POST",
            "/v1/license-keys/verify",
            body={
                "licenseKey": license_key,
                "productId": product_id,
                "machineId": machine_id,
            },
        )


class AsyncLicenseKeys(AsyncResource):
    """Asynchronous license-keys namespace."""

    async def create(
        self,
        *,
        organization_id: str,
        customer_id: str,
        product_id: str,
        entitlement_id: str,
        machine_limit: int,
    ) -> Any:
        """Issue a license key. ``POST /v1/license-keys``."""
        body = _create_body(
            organization_id=organization_id,
            customer_id=customer_id,
            product_id=product_id,
            entitlement_id=entitlement_id,
            machine_limit=machine_limit,
        )
        return await self._request("POST", "/v1/license-keys", body=body)

    async def verify(
        self, *, license_key: str, product_id: str, machine_id: str
    ) -> Any:
        """Verify a license key for a machine. ``POST /v1/license-keys/verify``."""
        return await self._request(
            "POST",
            "/v1/license-keys/verify",
            body={
                "licenseKey": license_key,
                "productId": product_id,
                "machineId": machine_id,
            },
        )
